### ISSUE TITLE
`this ref` struct extension shouldn't be considered overload for class type

### ISSUE THREAD

**Issue by benaadams** [dotnet/roslyn#25025]
```csharp
using System;
using System.Text;

public static class C 
{
    public static void Write<TBufferWriter, T>
        (this TBufferWriter bufferWriter, T[] source)
        where TBufferWriter : class, IBufferWriter<T>
        => WriteImpl<TBufferWriter, T>(ref bufferWriter, source);
    
    public static void Write<TBufferWriter, T>
        (this TBufferWriter bufferWriter, ReadOnlySpan<T> source)
        where TBufferWriter : class, IBufferWriter<T>
        => WriteImpl(ref bufferWriter, source);

    public static void Write<T, TBufferWriter>
        (this ref TBufferWriter bufferWriter, T[] source)
        where TBufferWriter : struct, IBufferWriter<T>
        => WriteImpl<TBufferWriter, T>(ref bufferWriter, source);
    
    public static void Write<T, TBufferWriter>
        (this ref TBufferWriter bufferWriter, ReadOnlySpan<T> source)
        where TBufferWriter : struct, IBufferWriter<T>
        => WriteImpl(ref bufferWriter, source);

    private static void WriteImpl<TBufferWriter, T>
        (ref TBufferWriter bufferWriter, ReadOnlySpan<T> source)
        where TBufferWriter : IBufferWriter<T>
    {
       // common implementation
    }
}

public interface IBufferWriter<T>
{
    void Advance(int count);
    Memory<T> GetMemory(int lengthHint = 0);
    Span<T> GetSpan(int lengthHint = 0);
}

public abstract class PipeWriter : IBufferWriter<byte>
{
    public abstract void Advance(int bytes);
    public abstract Memory<byte> GetMemory(int minimumLength = 0);
    public abstract Span<byte> GetSpan(int minimumLength = 0);
}

public struct StructWriter : IBufferWriter<byte>
{
    public void Advance(int bytes){}
    public Memory<byte> GetMemory(int minimumLength = 0) => default;
    public Span<byte> GetSpan(int minimumLength = 0) => default;
}
```
```csharp
public static void Main()
{
    PipeWriter buffer = null;
    buffer.Write(Encoding.UTF8.GetBytes("Hello World")); // ambiguous    
    
    StructWriter valueBuffer = default;
    valueBuffer.Write(Encoding.UTF8.GetBytes("Hello World")); // ambiguous    
}
```
Both calls are ambiguous even though only one can apply to each type due to `ref this` and `struct` and `class` constraints 
```csharp
The call is ambiguous between the following methods or properties: 
'C.Write<TBufferWriter, T>(TBufferWriter, T[])' and 
'C.Write<T, TBufferWriter>(ref TBufferWriter, T[])'
```

I assume the methods don't actually collide because adding types makes them happy
```csharp
public static void Main()
{
    PipeWriter buffer = null;
    buffer.Write<byte, PipeWriter>(Encoding.UTF8.GetBytes("Hello World"));   
    
    StructWriter valueBuffer = default;
    valueBuffer.Write<StructWriter, byte>(Encoding.UTF8.GetBytes("Hello World"));
}
```

Whereas reversing the type params causes both to error (for opposite reasons)
```csharp
public static void Main()
{
    PipeWriter buffer = null;
    buffer.Write<PipeWriter, byte>(Encoding.UTF8.GetBytes("Hello World"));   
    
    StructWriter valueBuffer = default;
    valueBuffer.Write<byte, StructWriter>(Encoding.UTF8.GetBytes("Hello World"));
}
```
```
error CS0453: The type 'PipeWriter' must be a non-nullable value type in order to use it as parameter 
'TBufferWriter' in the generic type or method 'C.Write<TBufferWriter, T>(ref TBufferWriter, T[])'

error CS0452: The type 'StructWriter' must be a reference type in order to use it as parameter 
'TBufferWriter' in the generic type or method 'C.Write<T, TBufferWriter>(TBufferWriter, T[])'
```

So is it really ambiguous? 

/cc @VSadov

**Comment by jcouv** [comment]
Tagging @OmarTawfik as well

**Comment by jcouv** [comment]
Also @gafter, I wonder if your recent change on improving the candidates for overload resolution (https://github.com/dotnet/roslyn/pull/24756) would affect this.

**Comment by VSadov** [comment]
https://github.com/dotnet/roslyn/pull/24756 should address this. The overload candidate that has no chance of succeeding due to constraints will be dropped from the list before it goes into resolution.  
 
It does not matter how the type parameter that does not satisfy constraints was specified/inferred - from ordinary parameter or from `this` of an extension.

**Comment by VSadov** [comment]
@benaadams - you can try your stuff in sharplab.io . Th e`features/compiler` branch has the feature.

https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGMD2BbADgSwBsIEAfAAQAYACcgRgG4BYAKCtroDoAVCADxjMWrcgGYOANloAmagGFqrAN6tqa2uPpTyAFmoB1BPhgQAPNwBCcMJASHjJADTVuAPlXrPAChgALfMguVjYk9iYI1MDWtmFOLgDaALrUyJiI6BAAlB6eagDuviQQQdGhRuHUIDLOAJLBMeUk5u4sueoAvK4GjTV4hOb1ZQ4Izm5eSGCRpXaNIylpCBmZQp45GpK0erED07FzbmvefgElITPDU2d7zgBK0AAmAPIAdoQAngDKuFDPzfPpWUO6gKRVODQuVXI0lqg3O4WaQLUnW6Dl6uEI4wgkyiV1mzlSAOWrDWYg2uhRJnMo1hexabTUPn8gQmYKGFRx4PCoyS/0WgNa9JBSFZcJIlRSMAQcHQMBhu1mCIFbWRsTR/Us8uGo1cmOxmq5vKWK3UJM0dG0W0aVJFtMR1EZJxZGtxFw5bLidygT1en2+vzchv59OoQuKzs5YqqyEl0tl1Dq+qaByVuRVPT6usuEbmBL5ROEKdwRgAbrBilpNhSIGqdi6Dcng5nw+6Im7RXNPd73l8fn9c0s7aGbbNxQm60m6eoVCn1AB6WfULDYbCYZ7UfB9CDYCDPGCwfCrtYAX1YJ4LpPwu5IYCgGXjNIVyennnJAEF7qXnhkvJeYIu0ru+aeAAslumAIG8fwAOIQDAoErhBP67tQxDPAA5n4AASv7UO01CUEB6g9v6XQwTAxFIX+qEYb42HIXhBFCGeIjiFAwDRggt5/lC1AAAr4LgEB7KOD7DKYwBvCYLTPuopJsRxXGVu+n7fjhEkmMghFqHJ7GSop8HgZB6kQKRsEGYhOHYJeG5wNgAAyO40bh+Faes8l6TK1DEeJkkmdQZEUZZ1nYLZDnoX4zmMaexJsOIHGxl5MYysJVRjtmPlScoppKR+PyqchxmaUoZ4vuI5lGb5plwWBFnIVZzw2fZjkRQxmS4V09xYlAcCEII2XecZVWBXVwWhc1viRW1yKdTePV9SwzFsLI9AAOxZSmpIVuS9AABxeNkKYyW0/GCcJbbOc8PWEMabRtpwsReAAol+mD3JeaGcAAqtwABiO2cGRFi+cgXgAESYRAhCEJgBjgYQ9yg5kyzUPO1BQNgwD4GhcBpIEJoFsGHxJTAwmloQcAQLCzkzd1vU3bkZMU7C92NE9L1vehX2/f9gPA2DENQzD+hwwjSMMCjC7o5j2O46sKZnkeQA===

**Comment by gafter** [comment]
Resolved in C# 7.3.
